README.md
**To make a build you must do the following:**

1) Change the current working directory to where your curent user has write permissions, for example `cd ~`
2) run `git clone --depth=1 --branch=main https://github.com/armbian/build`
3) Clone this repo with `git clone https://code.samourai.io/ronindojo/RoninOS.git`
4) run `mkdir -p build/userpatches/overlay/`
TODO: copy the lib.config file here to build/userpatches
5) run `cp -Rv RoninOS/customize-image.sh build/userpatches/customize-image.sh && cp -Rv RoninOS build/userpatches/overlay/`
6) change the working directory to the root of the build project with `cd build`
7) Start armbian build process by running the command `./compile.sh BOARD=rockpro64 BRANCH=current BUILD_DESKTOP=no BUILD_MINIMAL=yes KERNEL_CONFIGURE=no RELEASE=bullseye`
TODO: add the BSPFREEZE parameter to this compile.sh call
8) Let it run and relax. (Takes 45-90 min)
9) You can find the image files in the directory `output` that's in the root of the build project

**Resources:**

- https://docs.armbian.com/Developer-Guide_Build-Preparation/

**File structure:**

- _customize-image.sh:_ Customize the image during the building process. Setting username and password (random password). clone repos, modify tor and prep the UI. 

If any changes occur moving forward this repo will be updated. 
If you'd like to customize anything about the build, utilize the armbian guides and/or edit the `customize-image.sh` file.


CHANGELOG
Changelog

RoninOS v2.1.1 Change log
- Increased sudo timeout to 1 hour
- Simplified customize-image.sh image build script
- Reverted portion of refactoring that was necessary to remove the NOPASSWD property on the ronindojo user after initial boot installation procedure completes
- Credits: BTCxZelko, dammkewl


RoninOS v2.1.0 Change log

- Updated tor to 0.4.7
- Removed autologin regarding physical access
- Simplified ronin-setup.sh script (no need to change entries in the sudoers file)
- Cleanup (teardown .sh files)
- Improved installation service to wait for network interface to be up and have an uplink to the router
- Credits: dammkewl, BTCxZelko, s2l1, Numbers, KYC3, WittyBull


01-stage-ronin-os.sh

##
## This script will equip the Armbian build system with the RoninOS repo as an overlay
##


# Terminal colours
CDEF="\033[0m"
CGREEN="\033[0;32m"
CLGREEN="\033[1;32m"
CRED="\033[0;31m"

# Global vars
FILENAME=$(basename -- "$0";)



# Check for correct place of RoninOS repo
if [ ! -f ~/RoninOS/${FILENAME} ]; then
    echo " "
    echo -e "${FILENAME}: [${CRED} Failure ${CDEF}]"
    echo -e "RoninOS repo seems to not be present in the expected directory"
    echo -e "Please make sure you cloned it to [${CGREEN} ~/RoninOS ${CDEF}] to use the automated scripts"
    echo -e "If you want to operate manually, please refer to [${CGREEN} README.md ${CDEF}] and/or examine what the scripts intended to do"

    exit 0
fi

# Check for Armbian build dir present
if [ ! -f ~/build/compile.sh ]; then
    echo " "
    echo -e "${FILENAME}: [${CRED} Failure ${CDEF}]"
    echo -e "Armbian build directory seems to not be present as expected"
    echo -e "Please make sure you cloned it completely to [${CGREEN} ~/build ${CDEF}] to use the automated scripts"
    echo -e "If you want to operate manually, please refer to [${CGREEN} README.md ${CDEF}] and/or examine what the scripts intended to do"

    exit 0
fi



# If not already present, prepare 'overlay' dir for Armbian build system
mkdir -p ~/build/userpatches/overlay

# Copy RoninOS build script for Armbian build system to use
cp -fp ~/RoninOS/customize-image.sh ~/build/userpatches/customize-image.sh

# Delete potentially existing RoninOS dir in overlay
rm -Rfd ~/build/userpatches/overlay/RoninOS

# Copy RoninOS files to Armbian 'overlay' dir
cp -Rp ~/RoninOS ~/build/userpatches/overlay/

# Remove unnecessary .sh scripts to not confuse the user with doublet files
rm -f ~/build/userpatches/overlay/RoninOS/*.sh

# Output an encouraging success message
echo " "
echo -e "${FILENAME}: [${CLGREEN} Success ${CDEF}]"
echo -e "The RoninOS scripts are now prepared in [${CGREEN} ~/build/userpatches/overlay/ ${CDEF}]"
echo -e "If needed, make your edits to [${CGREEN} ~/build/userpatches/${CLGREEN}customize-image.sh ${CDEF}]"
echo -e "To make your build, run [ ${CGREEN}~/RoninOS/${CLGREEN}02-compile-ronin-os.sh ${CDEF}] and choose your target board."
echo -e "You will find your image then in [${CGREEN} ~/build/output/images/ ${CDEF}]"

# Unset executable to self, to not be able to accidentally mess up the files any more
chmod 664 ~/RoninOS/01-stage-ronin-os.sh
echo " "
echo -e "${CLGREEN}Note:${CDEF} This script is now set to be ${CGREEN}not executable${CDEF} (644) any more –"\
      "to not have you in the situation to accidentally mess up your staged files."\
      "However, should you be willing to use it again, simply command [${CGREEN} chmod 755 ${FILENAME} ${CDEF}] or execute [${CGREEN} 00-enable-staging.sh ${CDEF}] first."



overlays/RoninOS/usr/local/sbin/ronin-setup.sh
#!/bin/bash

regenerate_passwords_and_update_info_file(){
    # Generate a random password consisting of 16 characters using only numbers and letters.
    rootpwd="$(tr -dc 'a-zA-Z0-9' </dev/urandom | head -c'21')"
    roninpwd="$(tr -dc 'a-zA-Z0-9' </dev/urandom | head -c'21')"
    USER="ronindojo"


    # Change the root user's password.
    echo "Changing the root user's password..."
    sudo chpasswd <<<"root:$rootpwd"

    # Change the ronindojo user's password.
    echo "Changing the ronindojo user's password..."
    sudo chpasswd <<<"$USER:$roninpwd"

    # remove old info.json then create new with new passwords.
    rm -rf /home/ronindojo/.config/RoninDojo/info.json
    cat <<EOF >/home/"${USER}"/.config/RoninDojo/info.json
{"user":[{"name":"${USER}","password":"${roninpwd}"},{"name":"root","password":"${rootpwd}"}]}
EOF

    # add validation for that the setup was done.
    GENERATE_MESSAGE="Your password was randomly generated during System Setup."
    TIMESTAMP=$(date)
    cat <<EOF >/home/"${USER}"/.logs/pass_gen_timestamp.txt
$GENERATE_MESSAGE
Date and Time: $TIMESTAMP
EOF
}

cd "$HOME" || exit

# give time for Startup to finish before trying to update the repo. 
sleep 75s 

# Clone Repo
git clone -b master https://code.samourai.io/ronindojo/RoninDojo /home/ronindojo/RoninDojo
cd /home/ronindojo/RoninDojo

# Source files
. Scripts/defaults.sh
. Scripts/functions.sh

# Run main
if _main; then
    # regenerate the passwords to be random for each user.
    regenerate_passwords_and_update_info_file

    # Run system setup
    Scripts/Install/install-system-setup.sh system

    # Run RoninDojo install
    Scripts/Install/install-dojo.sh dojo

    # Restore getty
    sudo systemctl start ronin-post.service
    if ! systemctl is-active pm2-ronindojo.service; then
        sudo systemctl start pm2-ronindojo.service
    fi
    sudo systemctl disable ronin-setup.service
    sudo sed -i '/ronindojo/s/ALL) NOPASSWD:ALL/ALL) ALL/' /etc/sudoers
    touch /home/ronindojo/.logs/setup-complete
fi

customize-image.sh
#!/bin/bash

echo "add user roinindojo"
useradd -s /bin/bash -m -c "ronindojo" ronindojo -p rock
useradd -c "tor" tor && echo "ronindojo    ALL=(ALL) ALL" >> /etc/sudoers

#removes the first user login requirement with monitor and keyboard
rm /root/.not_logged_in_yet 

echo "set hostname"
hostname -b "ronindebian"

# RoninDojo part
TMPDIR=/var/tmp
USER="ronindojo"
#PASSWORD="password" # test purposes only
PASSWORD="$(tr -dc 'a-zA-Z0-9' </dev/urandom | head -c'21')"
#ROOTPASSWORD="password" ## for testing purposes only
ROOTPASSWORD="$(tr -dc 'a-zA-Z0-9' </dev/urandom | head -c'21')"
FULLNAME="RoninDojo"
TIMEZONE="UTC"
LOCALE="en_US.UTF-8"
HOSTNAME="RoninDojo"
KEYMAP="us"

_create_oem_install() {
    pam-auth-update --package	
    # Setting root password
    chpasswd <<<"root:$ROOTPASSWORD"

    # Adding user $USER
    useradd -m -G wheel,sys,audio,input,video,storage,lp,network,users,power -s /bin/bash "$USER" &>/dev/null

    # Set User and WorkingDirectory in ronin-setup.service unit file
    sed -i -e "s/User=.*$/User=${USER}/" \
        -e "s/WorkingDirectory=.*$/WorkingDirectory=\/home\/${USER}/" /usr/lib/systemd/system/ronin-setup.service

    # Setting full name to $FULLNAME
    chfn -f "$FULLNAME" "$USER" &>/dev/null

    # Setting password for $USER
    chpasswd <<<"$USER:$PASSWORD"

    # Save Linux user credentials for UI access
    mkdir -p /home/"${USER}"/.config/RoninDojo
    cat <<EOF >/home/"${USER}"/.config/RoninDojo/info.json
{"user":[{"name":"${USER}","password":"${PASSWORD}"},{"name":"root","password":"${ROOTPASSWORD}"}]}
EOF
    chown -R "${USER}":"${USER}" /home/"${USER}"/.config

    # Setting timezone to $TIMEZONE
    timedatectl set-timezone $TIMEZONE &>/dev/null
    timedatectl set-ntp true &>/dev/null

    # Generating $LOCALE locale
    sed -i "s/#$LOCALE/$LOCALE/" /etc/locale.gen &>/dev/null
    locale-gen &>/dev/null
    localectl set-locale $LOCALE &>/dev/null

    if [ -f /etc/sway/inputs/default-keyboard ]; then
        sed -i "s/us/$KEYMAP/" /etc/sway/inputs/default-keyboard

        if [ "$KEYMAP" = "uk" ]; then
            sed -i "s/uk/gb/" /etc/sway/inputs/default-keyboard
        fi
    fi

    # Setting hostname to $HOSTNAME
    hostnamectl set-hostname $HOSTNAME &>/dev/null

    # Resizing partition
    resize-fs &>/dev/null

    loadkeys "$KEYMAP"

    # Configuration complete. Cleaning up
    #rm /root/.bash_profile

    # Avahi setup
    sed -i 's/hosts: .*$/hosts: files mdns_minimal [NOTFOUND=return] resolve [!UNAVAIL=return] dns mdns/' /etc/nsswitch.conf
    sed -i 's/.*host-name=.*$/host-name=ronindojo/' /etc/avahi/avahi-daemon.conf
    if ! systemctl is-enabled --quiet avahi-daemon; then
        systemctl enable --quiet avahi-daemon
    fi

    # sshd setup
    sed -i -e "s/PermitRootLogin yes/#PermitRootLogin prohibit-password/" \
        -e "s/PermitEmptyPasswords yes/#PermitEmptyPasswords no/" /etc/ssh/sshd_config

    # Set sudo timeout to 1 hour
    sed -i '/env_reset/a Defaults\ttimestamp_timeout=60' /etc/sudoers

    # Enable passwordless sudo
    sed -i '/ronindojo/s/ALL) ALL/ALL) NOPASSWD:ALL/' /etc/sudoers # change to no password

    echo -e "domain .local\nnameserver 1.1.1.1\nnameserver 1.0.0.1" >> /etc/resolv.conf
    
    # Setup logs for outputs
    mkdir -p /home/ronindojo/.logs
    touch /home/ronindojo/.logs/setup.logs
    touch /home/ronindojo/.logs/post.logs
    chown -R ronindojo:ronindojo /home/ronindojo/.logs
}

_service_checks(){  
    if ! systemctl is-enabled tor.service; then
        systemctl enable tor.service
    fi

    if ! systemctl is-enabled --quiet avahi-daemon.service; then
        systemctl disable systemd-resolved.service &>/dev/null
        systemctl enable avahi-daemon.service
    fi

    if ! systemctl is-enabled motd.service; then
        systemctl enable motd.service
    fi
    
    if ! systemctl is-enabled ronin-setup.service; then
        systemctl enable ronin-setup.service
    fi

    if ! systemctl is-enabled ronin-post.service; then
        systemctl enable ronin-post.service
    fi
}

# Installs Nodejs, docker, docker-compose, and pm2. Clones the RoninDojo repo. This is needed due to some out dated packages in the default debian package manager.
_prep_install(){
    # install Nodejs
    apt-get install -y ca-certificates curl gnupg
    curl -fsSL https://deb.nodesource.com/gpgkey/nodesource-repo.gpg.key | gpg --dearmor -o /etc/apt/keyrings/nodesource.gpg
    NODE_MAJOR=20
    echo "deb [signed-by=/etc/apt/keyrings/nodesource.gpg] https://deb.nodesource.com/node_$NODE_MAJOR.x nodistro main" | tee /etc/apt/sources.list.d/nodesource.list
    apt-get update
    apt-get install -y nodejs

    # install docker
    mkdir -m 0755 -p /etc/apt/keyrings
    curl -fsSL https://download.docker.com/linux/debian/gpg | gpg --dearmor -o /etc/apt/keyrings/docker.gpg
    echo \
    "deb [arch=$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.gpg] https://download.docker.com/linux/debian \
    $(lsb_release -cs) stable" | tee /etc/apt/sources.list.d/docker.list > /dev/null

    apt-get update
    apt-get install -y docker-ce docker-ce-cli containerd.io docker-buildx-plugin docker-compose-plugin

    # install docker-compose
    curl -L https://github.com/docker/compose/releases/download/v2.23.0/docker-compose-linux-aarch64 -o /usr/bin/docker-compose
    chmod +x /usr/bin/docker-compose

    # install pm2 
    npm install pm2 -g
}

_ronin_ui_avahi_service() {
    if [ ! -f /etc/avahi/services/http.service ]; then
        tee "/etc/avahi/services/http.service" <<EOF >/dev/null
<?xml version="1.0" standalone='no'?><!--*-nxml-*-->
<!DOCTYPE service-group SYSTEM "avahi-service.dtd">
<!-- This advertises the RoninDojo vhost -->
<service-group>
 <name replace-wildcards="yes">%h Web Application</name>
  <service>
   <type>_http._tcp</type>
   <port>80</port>
  </service>
</service-group>
EOF
    fi

    sed -i 's/hosts: .*$/hosts: files mdns_minimal [NOTFOUND=return] resolve [!UNAVAIL=return] dns mdns/' /etc/nsswitch.conf

    if ! grep -q "host-name=ronindojo" /etc/avahi/avahi-daemon.conf; then
        sed -i 's/.*host-name=.*$/host-name=ronindojo/' /etc/avahi/avahi-daemon.conf
    fi

    if ! systemctl is-enabled --quiet avahi-daemon; then
        systemctl enable --quiet avahi-daemon
    fi

    return 0
}


_rand_passwd() {
    local _length
    _length="${1:-16}"

    tr -dc 'a-zA-Z0-9' </dev/urandom | head -c"${_length}"
}

# Install Ronin UI. This function is the same we utilize in the RoninDojo repo. Only modifying slightly since this runs during build and not organic setup.
_install_ronin_ui(){

    roninui_version_file="https://ronindojo.io/downloads/RoninUI/version.json"

    gui_api=$(_rand_passwd 69)
    gui_jwt=$(_rand_passwd 69)

    cd /home/ronindojo || exit

    npm i -g pnpm &>/dev/null

    #sudo npm install pm2 -g

    test -d /home/ronindojo/Ronin-UI || mkdir /home/ronindojo/Ronin-UI
    cd /home/ronindojo/Ronin-UI || exit

    wget -q "${roninui_version_file}" -O /tmp/version.json 2>/dev/null

    _file=$(jq -r .file /tmp/version.json)
    _shasum=$(jq -r .sha256 /tmp/version.json)

    wget -q https://ronindojo.io/downloads/RoninUI/"$_file" 2>/dev/null

    if ! echo "${_shasum} ${_file}" | sha256sum --check --status; then
        _bad_shasum=$(sha256sum ${_file})
        echo "Ronin UI archive verification failed! Valid sum is ${_shasum}, got ${_bad_shasum} instead..."
    fi
      
    tar xzf "$_file"

    rm "$_file" /tmp/version.json

    # Mark Ronin UI initialized if necessary
    if [ -e "${ronin_ui_init_file}" ]; then
        echo -e "{\"initialized\": true}\n" > ronin-ui.dat
    fi

    # Generate .env file
    echo "JWT_SECRET=$gui_jwt" > .env
    echo "NEXT_TELEMETRY_DISABLED=1" >> .env

    if [ "${roninui_version_staging}" = true ] ; then
        echo -e "VERSION_CHECK=staging\n" >> .env
    fi

    _ronin_ui_avahi_service

    chown -R ronindojo:ronindojo /home/ronindojo/Ronin-UI
}

# The debian default was incompatible with our setup. This sets tor to match RoninDojo requirements and removes the debian variants.
_prep_tor(){
	mkdir -p /mnt/usb/tor
	chown -R tor:tor /mnt/usb/tor
	sed -i '$a\User tor\nDataDirectory /mnt/usb/tor' /etc/tor/torrc
    sed -i '$ a\
HiddenServiceDir /mnt/usb/tor/hidden_service_ronin_backend/\
HiddenServiceVersion 3\
HiddenServicePort 80 127.0.0.1:8470\
' /etc/tor/torrc

    cp -Rv /tmp/RoninOS/overlays/RoninOS/example.tor.service /usr/lib/systemd/system/tor.service
    rm -rf /usr/lib/systemd/system/tor@* #remove unnecessary debian installed services
}

# This installs all required packages needed for RoninDojo. Clones the RoninOS repo so it can be copied to appropriate locations. Then runs all the functions defined above.
main(){
    # install dependencies
    apt-get update
    apt-get install -y man-db git avahi-daemon nginx openjdk-11-jdk fail2ban net-tools htop unzip wget ufw rsync jq python3 python3-pip pipenv gdisk gcc curl apparmor ca-certificates gnupg lsb-release nvme-cli 
    apt-get install -y tor/bullseye-backports #install 0.4.7.x tor
    # clone the original RoninOS
    git clone https://code.samourai.io/ronindojo/RoninOS.git /tmp/RoninOS

    cp -Rv /tmp/RoninOS/overlays/RoninOS/usr/* /usr/
    cp -Rv /tmp/RoninOS/overlays/RoninOS/etc/* /etc/
    ### sanity check ###
    # TODO: Remove this after successful runs.
    if [ ! -f /usr/lib/systemd/system/ronin-setup.service ]; then
        echo "ronin-setup.service is missing..."
        echo "Still broken.. exiting"
        exit 1;
    else 
        echo "Setup service is PRESENT! Keep going!"
        _create_oem_install
        _prep_install
        _prep_tor
        usermod -aG pm2 ronindojo
        mkdir -p /usr/share/nginx/logs
        rm -rf /etc/nginx/sites-enabled/default
        _install_ronin_ui
        usermod -aG docker ronindojo
        systemctl enable oem-boot.service
        _service_checks
        echo "Setup is complete"
    fi
}

# Run main setup function.
main
